Introduced `raft::state_machine` initial recovery policy

mmaslankaprv opened the pull request:
Introduced the `stm_initial_recovery_policy`. The policy is there for
the state machine implementer to decide how to handle the situation when
a state machine is crated for the first time. The state machine can
choose from two options:

`read_everything` - the read everything policy simply creates an stm and
apply each batch of the whole locally available log. The application is
being done in the `background` apply fiber of state machine manager
until the stm is fully caught up. (This was the default behavior)

`skip_to_end` - this policy allows the state machine not to apply the
whole log when it is created for the first time. The skip to the end
policy starts state machine initial recovery from the very specific
offset. The offset is established as the next to highest last applied
offset returned from all the state machines created on the same
partition previously.

Why we can not skip to the log end offset with `skip_to_end` ?

We can not skip to the end offset of the log as it may have not been
committed and therefore can be truncated. In this situation the stm
could loose some applies as its `_next` wouldn't change with the
truncation and only batches with offset greater than `_next` would be
applied to the state machine. Using last applied offset from other state
machines eliminates this problem as last applied offset must have been
committed before to be applied.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Improvements
- give state machine implementer a control over the initial recovery policy

mmaslankaprv committed:
r/state_machine: introduced initial recovery policy

Introduced the `stm_initial_recovery_policy`. The policy is there for
the state machine implementer to decide how to handle the situation when
a state machine is crated for the first time. The state machine can
choose from two options:

`read_everything` - the read everything policy simply creates an stm and
apply each batch of the whole locally available log. The application is
being done in the `background` apply fiber of state machine manager
until the stm is fully caught up. (This was the default behavior)

`skip_to_end` - this policy allows the state machine not to apply the
whole log when it is created for the first time. The skip to the end
policy starts state machine initial recovery from the very specific
offset. The offset is established as the next to highest last applied
offset returned from all the state machines created on the same
partition previously.

Why we can not skip to the log end offset with `skip_to_end` ?

We can not skip to the end offset of the log as it may have not been
committed and therefore can be truncated. In this situation the stm
could loose some applies as its `_next` wouldn't change with the
truncation and only batches with offset greater than `_next` would be
applied to the state machine. Using last applied offset from other state
machines eliminates this problem as last applied offset must have been
committed before to be applied.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/tests: added initial recovery policy tests

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/translation_stm: use skip_to_end initial recovery policy

Using skip_to_end initial recovery policy in
`datalake::translation::state_machine`. The state machine is only
applying datalake state machine batches therefore it doesn't have to
read the whole log during initial recovery as there are no datalake
state machine batches in the earlier state.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/parittion_properties_stm: use skip_to_end initial recovery policy

If the `partition_properties_stm` did not exists on the partition it
doesn't have to have the whole log applied.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
tests/redpanda: added method returning estimate of total bytes read

The estimate may be helpful to validate the amount of data read during
recovery

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
tests: added test verifying initial recovery policy for translation stm

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
tests/upgrade: added test verifying total bytes read during recovery

Added test verifying if there is an increase in the total bytes that are
read during the first startup of Redpanda after upgrade. The increased
read size may cause high disk and cpu usage in cluster with lots of data
retained locally.

The test makes sure that the total amount of bytes read after upgrade is
not increased more than 10%.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64686-->
<!--eb02d9b0-6eea-4086-8505-ba662b8f08e4-->
<!--0184ea0d-ca00-49d7-9ee0-a3892faa1c81-->
<!--title-->
## Retry command for [Build#64686](https://buildkite.com/redpanda/redpanda/builds/64686)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b5c2cf5d-c43f-4a16-9879-49543f55341f-->
<!--TEST RESULTS-->
<!--d2e9642e-b8d0-4cc6-b82a-a8a79af9984a-->
<!--TEST RESULTS-->
<!--a7b65aef-700e-457f-b809-d1a2b5c0f83a-->
<!--TEST RESULTS-->
<!--3802bc81-3986-4984-a87b-c8df47170837-->
<!--TEST RESULTS-->
<!--2886ee38-977b-45df-8eb0-1ca34a5fd00c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64686">build#64686</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_storage_e2e_rpfixture.gtest_storage_e2e_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64686#0196445e-469a-4432-aa0a-7b498d836873 | FLAKY | 1/2 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64686#019644a7-f469-45c7-86dc-3ac491c2f51f | FLAKY | 15/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64686#019644a7-f469-4322-83dc-93925baf7f4f | FLAKY | 13/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64754">build#64754</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64754#019648f9-dc9a-43f4-906d-4c35ac518286 | FLAKY | 1/2 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/64754#01964954-37a0-4fe0-9bb5-1ac4ce17ae99 | FLAKY | 19/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_higher_level_migration_api | ducktape | https://buildkite.com/redpanda/redpanda/builds/64754#01964951-990d-4a10-b95c-fdb51488c8e9 | FLAKY | 19/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64754#01964951-990b-4030-81d7-9cd672caa46a | FLAKY | 16/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64754#01964951-990e-4d45-91fb-012dac22e991 | FLAKY | 12/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64836">build#64836</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64836#01965c7a-62fc-438c-8d2f-467e3a4023dc | FLAKY | 19/21 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64836#01965c68-8c1f-4cb0-9554-dc2a8ec6f005 | FLAKY | 19/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64836#01965c68-8c1f-4cb0-9554-dc2a8ec6f005 | FLAKY | 15/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64836#01965c68-8c1e-4c37-b740-03dc756792f6 | FLAKY | 13/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64855">build#64855</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64855#01965dab-1a6d-462e-a1d7-7ae8e2276772 | FLAKY | 1/2 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/64855#01965e03-30d7-4886-848d-da604857f796 | FLAKY | 20/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64855#01965e03-30d8-4301-b47c-7ceaf20f1d50 | FLAKY | 16/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64855#01965e03-30d5-4e71-9f16-b49e874c1a6b | FLAKY | 13/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64948">build#64948</a></summary>


| test_id                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64948#01966340-9b12-4e78-befe-07833374ca3a | FLAKY         | 9/21     |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64754-->
<!--f56788f4-2add-4de3-83ab-3d65ac053693-->
<!--2dcd0cdd-9930-446e-8511-9ec83f4031b6-->
<!--e9cd433e-a17b-4142-b832-f1c392e46539-->
<!--56958bb4-fd6b-4ad4-bdd5-8b9f6f350a4e-->
<!--title-->
## Retry command for [Build#64754](https://buildkite.com/redpanda/redpanda/builds/64754)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_higher_level_migration_api
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/controller_log_limiting_test.py::ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64836-->
<!--0ec1e5ce-c9e1-49f4-8d24-308f403f9dbd-->
<!--8ba9e5b5-b1d3-4364-a139-93f1de0e6810-->
<!--0fd5f9d6-5118-41c3-a355-ebab792a2e5a-->
<!--title-->
## Retry command for [Build#64836](https://buildkite.com/redpanda/redpanda/builds/64836)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":false}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

bashtanov commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -952,3 +956,100 @@ def test_rest_catalog_metrics(self, cloud_storage_type):
+
+            # produce ~60 MiB to the topic
+            dl.produce_to_topic(topic.name, 1024, 120 * 1024)
+            # waif for a while for the local snapshot to be taken
```
typo:wait

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64855-->
<!--bed21451-b110-40d3-9a6a-de6753f8bcf0-->
<!--bbacbf3d-9932-4606-b05e-db4553e00b73-->
<!--c3a32f25-6e6c-4abf-9455-3e123ff9688f-->
<!--title-->
## Retry command for [Build#64855](https://buildkite.com/redpanda/redpanda/builds/64855)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/controller_log_limiting_test.py::ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
```
<!--/body-->

ztlpn commented on src/v/raft/state_machine_base.h:
```diff
@@ -97,6 +119,28 @@ class state_machine_base {
+     * machine
+     */
+    virtual stm_initial_recovery_policy get_initial_recovery_policy() const {
+        return stm_initial_recovery_policy::read_everything;
```
Not sure about this being a good default, for new state machines this is probably should be `skip_to_end`, but discoverability is low - it is easy to forget to change it. Should this be a pure virtual instead to force the stm authors to make a choice? (could be a followup)

ztlpn commented on tests/rptest/services/redpanda.py:
```diff
@@ -5493,6 +5493,22 @@ def estimate_bytes_written(self):
         else:
             return None
 
+    def estimate_bytes_read(self):
```
nit: since this a common helper, a bit more descriptive name (like `estimate_disk_bytes_read` to distinguish from e.g. kafka bytes read) would be nice.

ztlpn reviewed (dismissed):
LGTM, left a couple of comments that can be addressed as follow ups

andrwng commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -952,3 +956,100 @@ def test_rest_catalog_metrics(self, cloud_storage_type):
+            bytes_read_before = self.redpanda.estimate_bytes_read()
+
+            # enable iceberg, this will restart the cluster
+            dl.redpanda.set_cluster_config({"iceberg_enabled": True},
```
Now that we have this new mechanism to avoid extra reads for Iceberg, should we remove the condition in the STM factory that disables the STM iceberg_enabled is false? The new approach seems more robust and doesn't suffer from the lack of determinism that the cluster property has

mmaslankaprv replied:
let's do it in the follow up PR, ok ?

andrwng commented on src/v/raft/persisted_stm.h:
```diff
@@ -160,6 +160,13 @@ concept supported_stm_snapshot = requires(T s, stm_snapshot&& snapshot) {
+ * !!!! WARNING !!!!
+ * Removal of the persisted stm local snapshot.
+ *
+ * Removing persistent stm snapshot will trigger its initial recovery policy to
+ * be applied. This may lead to skipping some of the batches that may be
+ * important when the stm uses `skip_to_end` recovery policy.
```
I don't have a good/strong suggestion here, but it might be nice if we had a place to store the existence of STMs (like some kind of stm manager snapshot). Then we could tell if this were indeed the first time bootstrapping or if the local snapshot were deleted.

mmaslankaprv replied:
I actually pursued this idea, and change the way initial recovery information is stored

mmaslankaprv committed:
r/state_machine_base: made `get_initial_recovery_policy` abstract

Made the `get_initial_recovery_policy` method abstract to force state
machine implementation to choose the right policy.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -219,6 +223,8 @@ ss::future<> state_machine_manager::start() {
       _log.debug,
       "started state machine manager with initial next offset: {}",
       _next);
+    // it safe to update next here as the apply loop didn't started yet.
```
nit grammar: "didn't start" or "hasn't started"

ztlpn commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -1058,3 +1062,100 @@ def test_rest_catalog_metrics(self, cloud_storage_type):
+
+            DefaultClient(dl.redpanda).create_topic(topic)
+
+            # produce ~60 MiB to the topic
```
nit: should this be 120? :)

ztlpn commented on src/v/raft/state_machine_manager.h:
```diff
@@ -220,6 +256,7 @@ class state_machine_manager final {
     ss::abort_source _as;
     ss::scheduling_group _apply_sg;
     snapshot_at_offset_supported _supports_snapshot_at_offset{true};
+    storage::simple_snapshot_manager _snapshot_mgr;
```
Nit: a more specific variable name would be nice. Or is the idea to maybe put other things into this snapshot in the future?

ztlpn commented on src/v/raft/state_machine_manager.h:
```diff
@@ -139,6 +139,31 @@ class state_machine_manager final {
+     * that require that.
+     */
+    struct initial_recovery_snapshot
+      : serde::envelope<
```
Is this data automatically checksummed by the snapshot manager or do we need to use `checksum_envelope`?

mmaslankaprv replied:
yes, snapshot is checksummed

ztlpn replied:
Can you point me to where it happens? AFAICS `writer.output()` is just a regular file stream so we don't calculate any checksum when we write out the snapshot data and don't write any kind of footer.

ztlpn commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
+        }
+    }
+
+    // clean up offsets from state machines that are not longer present in the
```
This seems correct for stms that are deleted from the codebase, but not very intuitive behavior for stms that are removed and then re-added due to some configuration changes, is this the behavior that we want for them?

mmaslankaprv replied:
i honestly think this should be the default, i.e. when STM is re added to the partition it should not bother about the pervious state but only look at the current one

ztlpn commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
+          "Applying '{}' initial recovery policy for '{}' state machine",
+          policy,
+          name);
+        if (policy == stm_initial_recovery_policy::read_everything) {
```
nit: looks like now for every policy we do something so a switch looks more appropriate

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
+            entry->stm->set_next(std::max(it->second, entry->stm->next()));
+            continue;
+        }
+        // Stm needs initial recovery
```
Even if it has a local snapshot and it's the first time we are running the new Redpanda version?

mmaslankaprv replied:
addressed that

bashtanov commented on src/v/raft/state_machine_manager.cc:
```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
+              "skipping initial recovery for {} state machine as it was "
+              "recovered from snapshot",
+              name);
+            entry->stm->set_next(std::max(it->second, entry->stm->next()));
```
I'm not sure why we take the max here. STM next is either 0 (uninitialized) or comes from a local snapshot. Do we assume it cannot be behind the value from STMM snapshot? Why is that?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64929-->
<!--88cd7c21-ed1e-4b5a-84c5-22058109e044-->
<!--a0a9bc72-6bf0-458f-a93e-a11a0a11154f-->
<!--title-->
## Retry command for [Build#64929](https://buildkite.com/redpanda/redpanda/builds/64929)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64948-->
<!--eef132fb-3d3f-46a4-b541-05ee5af8a912-->
<!--title-->
## Retry command for [Build#64948](https://buildkite.com/redpanda/redpanda/builds/64948)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
```
<!--/body-->

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25840-v24.3.x-784 remotes/upstream/v24.3.x
git cherry-pick -x e29eca5a29 d90da97550 cf03f404af 2f2496db37 990611ae4e a1eb2daa39 f0baf41313 d9f58b059a
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14626564912).
